File: Data Structures/Graphs/graphs.cpp
/**
========================================================================
@file graphs.cpp

Big O
------

Space Complexity:
    Let 
        V := number of vertices
        E := number of edges

    Adjacent  matrix:   O(|V|^2)
    Adjacency list:     O(|V|+|E|)

Time Complexity:
    Add Vertex:
        Adjacent  matrix:   O(|V|^2)
        Adjacency list:     O(1)
    
    Add Edge:
        Adjacent  matrix:   O(1)
        Adjacency list:     O(1)

    Remove Edge:
        Adjacent  matrix:   O(1)
        Adjacency list:     O(1)

    Remove Vertex:
        Adjacent  matrix:   O(|V|^2)
        Adjacency list:     O(V)

: zach wolpe
: zach.wolpe@medibio.com.au
: 04 July 2023
========================================================================
*/

#include <iostream>
#include <unordered_map>
#include <unordered_set>

class Graph {
private:
    std::unordered_map< std::string, std::unordered_set<std::string> > adjList;

public:
    void printGraph() {
        std::unordered_map< std::string, std::unordered_set<std::string> >::iterator kvPair = adjList.begin();
        while (kvPair != adjList.end()) {
            std::cout << kvPair->first << ": [ ";
            std::unordered_set<std::string>::iterator edge = kvPair->second.begin();
            while (edge != kvPair->second.end()) {
                std::cout << edge->data() << " "; // view edges
                edge++;
            }
            std::cout << "]" << std::endl;
            kvPair++;
        }
    }
    
    bool addVertex(std::string vertex) {
        if (adjList.count(vertex) == 0) {
            adjList[vertex];
            return true;
        }
        return false;
    };

    bool addEdge(std::string vertex1, std::string vertex2) {
        if (adjList.count(vertex1) != 0 && adjList.count(vertex2) != 0) {
            adjList.at(vertex1).insert(vertex2);
            adjList.at(vertex2).insert(vertex1);
            return true;
        }
        return false;
    };

    bool removeEdge(std::string vertex1, std::string vertex2) {
        if (adjList.count(vertex1) != 0 && adjList.count(vertex2) != 0) {
            adjList.at(vertex1).erase(vertex2);
            adjList.at(vertex2).erase(vertex1);
            return true;
        }
        return false;
    };

    bool removeVertex(std::string vertex) {
        if (adjList.count(vertex) == 0) return false;
        for (auto otherVertex : adjList.at(vertex)) {
            adjList.at(otherVertex).erase(vertex);
        }
        adjList.erase(vertex);
        return true;
    };
};

int main() {
    std::cout << "\n\n< DevTools :/ " << std::endl;

    Graph* graph = new Graph();
    graph->addVertex("A");
    graph->addVertex("B");
    std::cout << "Graph:\n";
    graph->printGraph();

    std::cout << "........................." << std::endl; 
    std::cout << "Graph before addEdge(): \n";
    graph->printGraph();
    graph->addEdge("A", "B");
    std::cout << "\n\nGraph after addEdge(): \n";
    graph->printGraph();

    return 0;
}

File: Data Structures/Graphs/graphs.py
"""
========================================================================
@file graphs.cpp

Big O
------

Space Complexity:
    Let 
        V := number of vertices
        E := number of edges

    Adjacent  matrix:   O(|V|^2)
    Adjacency list:     O(|V|+|E|)

Time Complexity:
    Add Vertex:
        Adjacent  matrix:   O(|V|^2)
        Adjacency list:     O(1)
    
    Add Edge:
        Adjacent  matrix:   O(1)
        Adjacency list:     O(1)

    Remove Edge:
        Adjacent  matrix:   O(1)
        Adjacency list:     O(1)

    Remove Vertex:
        Adjacent  matrix:   O(|V|^2)
        Adjacency list:     O(V)

: zach wolpe
: zach.wolpe@medibio.com.au
: 15 July 2023
========================================================================
"""

from abc import ABC

class Graph:
    def __init__(self):
        self.adj_list = {}

    def printGraph(self):
        for vertex in self.adj_list:
            print(vertex, ':', self.adj_list[vertex])

    def addVertex(self, vertex:str):
        if vertex not in self.adj_list.keys():
            self.adj_list[vertex] = []
            return True
        return False
    
    def addEdge(self, vertex1, vertex2):
        if (vertex1 in self.adj_list.keys() and vertex2 in self.adj_list.keys()):
            self.adj_list[vertex1].append(vertex2)
            self.adj_list[vertex2].append(vertex1)
            return True
        return False
    
    def removeEdge(self, vertex1, vertex2):
        if (vertex1 in self.adj_list.keys() and vertex2 in self.adj_list.keys()):
            try:
                self.adj_list[vertex1].remove(vertex2)
                self.adj_list[vertex2].remove(vertex1)
                return True
            except ValueError:
                pass
        return False
            
    def removeVertex(self, vertex):
        if vertex in self.adj_list.keys():
            for other_vector in self.adj_list[vertex]:
                self.adj_list[other_vector].remove(vertex)
            del self.adj_list[vertex]
        return False

    def remove_vertex(self, vertex):
        if vertex in self.adj_list.keys():
            for other_vertex in self.adj_list[vertex]:
                self.adj_list[other_vertex].remove(vertex)
            del self.adj_list[vertex]
            return True
        return False

        



if __name__ == '__main__':

    my_graph = Graph()
    my_graph.add_vertex('A')
    my_graph.add_vertex('B')
    my_graph.add_vertex('C')
    my_graph.add_vertex('D')

    my_graph.add_edge('A','B')
    my_graph.add_edge('A','C')
    my_graph.add_edge('A','D')
    my_graph.add_edge('B','D')
    my_graph.add_edge('C','D')


    print('Graph before remove_vertex():')
    my_graph.print_graph()


    my_graph.remove_vertex('D')


    print('\nGraph after remove_vertex():')
    my_graph.print_graph()



    """
    EXPECTED OUTPUT:
    ----------------
        Graph before remove_vertex():
        A : ['B', 'C', 'D']
        B : ['A', 'D']
        C : ['A', 'D']
        D : ['A', 'B', 'C']

        Graph after remove_vertex():
        A : ['B', 'C']
        B : ['A']
        C : ['A']

    """